source/NVDAHelper.py

```diff
@@ -63,31 +74,118 @@ def nvdaController_speakText(text):
 	focus=api.getFocusObject()
 	if focus.sleepMode==focus.SLEEP_FULL:
 		return -1
-	import queueHandler
 	import speech
 	queueHandler.queueFunction(queueHandler.eventQueue,speech.speakText,text)
-	return 0
+	return SystemErrorCodes.SUCCESS
+
+
+# C901 'nvdaController_speakSsml' is too complex
+# Note: when working on nvdaController_speakSsml, look for opportunities to simplify
+# and move logic out into smaller helper functions.
+@WINFUNCTYPE(c_long, c_wchar_p, c_int, c_int, c_bool)
+def nvdaController_speakSsml(  # noqa: C901
+		ssml: str,
+		symbolLevel: "SymbolLevel",
+		priority: "SpeechPriority",
+		asynchronous: bool,
+) -> SystemErrorCodes:
+	focus = api.getFocusObject()
+	if focus.sleepMode == focus.SLEEP_FULL:
+		return SystemErrorCodes.ACCESS_DENIED
+
+	import speech
+	from characterProcessing import SymbolLevel
+	from speech.priorities import SpeechPriority
+	from speech.speech import _getSpeakSsmlSpeech
+
+	try:
+		symbolLevel = SymbolLevel(symbolLevel)
+	except ValueError:
+		log.exception("Invalid symbolLevel")
+		return SystemErrorCodes.INVALID_PARAMETER
+
+	try:
+		priority = SpeechPriority(priority)
+	except ValueError:
+		log.exception("Invalid SpeechPriority")
+		return SystemErrorCodes.INVALID_PARAMETER
+
+	prefixSpeechCommand = None
+	markCallable = None
+	if not asynchronous:

```

Comment:
Am I write in understanding that if the function is called with asynchronous as True, the mark callback will not be ever fired?

Reply:
This is correct. As per the [docs about callbacks](https://learn.microsoft.com/en-us/windows/win32/midl/callback):

> The callback function ... lets the server query the client for information in the context of the original call. Callbacks are special cases of remote calls that execute as part of a single thread. A callback is issued in the context of a remote call. Any remote procedure defined as part of the same interface as the static callback function can call the callback function.

So in other words, a callback can only be called during a call. Calling the callback as part of an async call that was already completed wouldn't work.